### ISSUE TITLE
Use MyPy support for SQLAlchemy

### ISSUE THREAD

**Issue by kaxil** [apache/airflow#20484]
### Body

https://docs.sqlalchemy.org/en/14/orm/extensions/mypy.html describes how type checking can be used with SQLAlchemy.

This is possible after we bump the SQLAlchemy version to 1.4+

**Comment by uranusjr** [comment]
Except we're still on 1.3 😔

**Comment by kaxil** [comment]
Yeah :( Which dep is causing us to stay on <1.4, I can't remember anymore

**Comment by kaxil** [comment]
https://github.com/apache/airflow/pull/16630

Flask-Appbuilder ... :(

**Comment by potiuk** [comment]
Ehhhh

**Comment by kaxil** [comment]
Actually, it looks like it is because of https://github.com/pallets/flask-sqlalchemy/pull/1001 that FAB has to still use <1.4. FAB Issue: https://github.com/dpgaspar/Flask-AppBuilder/issues/1710

**Comment by potiuk** [comment]
![dependency](https://user-images.githubusercontent.com/595491/147264001-06a48702-c5cb-414b-80c1-dc6c7d581b5a.png)

**Comment by kaxil** [comment]
Very apt XKCD 😄

**Comment by bmckallagat-os** [comment]
@kaxil I noticed [they deleted your comment](https://github.com/pallets/flask-sqlalchemy/pull/1001#event-5810515035) and locked the convo... petition to start an opensource war?

**Comment by potiuk** [comment]
> @kaxil I noticed [they deleted your comment](https://github.com/pallets/flask-sqlalchemy/pull/1001#event-5810515035) and locked the convo... petition to start an opensource war?

I started a few already, some of them with success .... some of them ... not yet. Maybe time for another one ;)

**Comment by bmckallagat-os** [comment]
> > @kaxil I noticed [they deleted your comment](https://github.com/pallets/flask-sqlalchemy/pull/1001#event-5810515035) and locked the convo... petition to start an opensource war?
> 
> I started a few already, some of them with success .... some of them ... not yet. Maybe time for another one ;)

Please, use that Apache clout 🤣

**Comment by uranusjr** [comment]
I took a look at the code path and tried an alternative fix: dpgaspar/Flask-AppBuilder#1783

**Comment by boredland** [comment]
so isn't this done now via https://github.com/dpgaspar/Flask-AppBuilder/pull/1786 ?

**Comment by potiuk** [comment]
It's in https://pypi.org/project/Flask-AppBuilder/3.4.4rc1/ - so as soon as it's released we will be able to switch to it.

**Comment by boredland** [comment]
perhaps I am misunderstanding something, but to me it looks like somewhere (didn't find where) there seems to be a hardcoded prefix 'postgres://' for connections. that surely will break after the 1.4. upgrade and would have to be 'postgresql://': 
![image](https://user-images.githubusercontent.com/4662748/150556813-27a36ba7-7baf-408d-97cb-575a14d2d1e5.png)

**Comment by zzzeek** [comment]
hi just an FYI, the SQLAlchemy mypy plugin is not going anywhere but in the 2.0 series of SQLAlchemy we hope to have a completely plugin-less approach to typing in place and the mypy plugin itself should become legacy.  bugs within the mypy plugin that don't have an easy fix we will likely just "wontfix".   so support the plugin sure but don't get too reliant on it.

**Comment by potiuk** [comment]
> perhaps I am misunderstanding something, but to me it looks like somewhere (didn't find where) there seems to be a hardcoded prefix 'postgres://' for connections. that surely will break after the 1.4. upgrade and would have to be 'postgresql://': ![image](https://user-images.githubusercontent.com/4662748/150556813-27a36ba7-7baf-408d-97cb-575a14d2d1e5.png)

Not really this is "Airflow" connection to connect to Postgres via "postgres provider" - it's not even an SQLAlchemy connection - we are using DBApi and Postgres Provider exposes plain old "SQL" interface to Postgres Provider (It would be next to impossible to have generic SQLAlchemy ORM interface).

**Comment by saulbein** [comment]
[dpgaspar/Flask-AppBuilder/releases/tag/v3.4.4](https://github.com/dpgaspar/Flask-AppBuilder/releases/tag/v3.4.4) the fix has now been released - it would be nice to be able to update sqlalchemy to 1.4 since we have some other dependencies that we would like to update and they require 1.4.

**Comment by potiuk** [comment]
@saulbein  - maybe you would like to add support for MyPY - seems that most of the problems with SQLAlchemy 1.4 have been addresse,d you could take care about it ?

**Comment by bmckallagat-os** [comment]
@potiuk Are there any other related issues besides https://github.com/apache/airflow/pull/20874 https://github.com/apache/airflow/pull/19294 https://github.com/apache/airflow/pull/21294 https://github.com/apache/airflow/pull/21296 https://github.com/apache/airflow/pull/21272?

Happy to help but those seem to be all resolved or taken.

**Comment by potiuk** [comment]
go ahead. There is still an issue with going beyond 1.9 (but It should not clash with mypy support)

**Comment by saulbein** [comment]
@potiuk I would like to help, but sadly I don't know much (if at all) about Flask-AppBuilder and can't allocate enough time to learn with other work :disappointed:. I'll see if I can at least contribute some typing.

**Comment by bmckallagat-os** [comment]
@potiuk I apologize, I just saw this now. I'm going on vacation until the 7th and I will try to tackle this after that time. If anyone wants to assign to themselves, please feel free. Otherwise I will make sure to get this done then.

**Comment by bmckallagat-os** [comment]
@potiuk Finally getting back to this. Just wanted to note that the docs listed in the description state that there will be native type checking without the use of the plugin when 2.0 is released and it is now [considered a legacy library](https://docs.sqlalchemy.org/en/14/orm/extensions/mypy.html). 

I'll just add that change however and add a note in the setup.cfg.

**Comment by potiuk** [comment]
Cool

**Comment by eladkal** [comment]
https://docs.sqlalchemy.org/en/20/orm/extensions/mypy.html

>However, SQLAlchemy 2.0, when released, will feature an [all new typing system](https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#whatsnew-20-orm-declarative-typing) for ORM Declarative models that removes the need for the Mypy plugin and delivers much more consistent behavior with generally superior capabilities.

I guess we should wait till after we bump to 2.0 (after we completely remove 1.4)? https://github.com/apache/airflow/issues/28723

**Comment by potiuk** [comment]
@bmckallagat-os We are unassigning you from this issue as part of our updated [assignment policy](https://github.com/apache/airflow/blob/main/contributing-docs/04_how_to_contribute.rst#contribute-code-changes).

This is not meant to discourage your contribution — quite the opposite! You are still very welcome to work on this issue and submit a PR for it. Simply comment that you are working on it and open a PR when ready.

We found that formal assignments were not working well, as they often prevented others from contributing when the assignee was not actively working on the issue.
